lib/fluent/plugin/out_google_cloud.rb

```diff
@@ -374,31 +373,28 @@ def write(chunk)
             if record.key?('kubernetes')
               handle_container_metadata(record, entry)
             end
-            # Save the timestamp if available
-            timestamp = record.key?('time') ? record['time'] : nil
-            # If the log from the user container is json, we want to export it
-            # as a structured log. Now that we've pulled out all the
-            # container-specific metadata from the record, we can replace the
-            # record with the json that the user logged.
-            # To save CPU in the common case of unstructured logs, only check if
-            # the contents are parsable as json for the first entry of each
-            # batch.
-            if is_container_json.nil? && record.key?('log')
-              record_json = parse_json_or_nil(record['log'])
-              if record_json.nil?
-                is_container_json = false
-              else
-                record = record_json
-                is_container_json = true
-              end
-            elsif is_container_json && record.key?('log')
-              record_json = parse_json_or_nil(record['log'])
-              record = record_json unless record_json.nil?
-            end
-            # Restore timestamp if necessary
-            unless record.key?('time') || timestamp.nil?
-              record['time'] = timestamp
-            end
+          end
+
+          # Save the timestamp if available, then clear it out to allow for
+          # determining whether we should parse the log or message field.
+          timestamp = record.key?('time') ? record['time'] : nil
+          record.delete('time')
+          # If the log is json, we want to export it as a structured log unless
+          # there is additional metadata that would be lost.
+          is_json = false
+          if record.length == 1 && record.key?('log')
+            record_json = parse_json_or_nil(record['log'])
```

Comment:
I disagree: I believe encouraging/nudging users towards using structured
logs data is a good thing, and would see parsing JSON as a key way to do
that. It's a benign automatism because it doesn't lead to data loss, and
because it is easy to understand what happens. Not parsing JSON means that
you ask users to perform an action (changing configuration), which makes it
less likely that logs data becomes structured data.

On Thu, 16 Jun 2016, 20:21 Todd Derr, <notifications@github.com> wrote:

> In lib/fluent/plugin/out_google_cloud.rb
> <https://github.com/GoogleCloudPlatform/fluent-plugin-google-cloud/pull/77#discussion_r67397835>
> :
>
> > -            end
> > -            # Restore timestamp if necessary
> > -            unless record.key?('time') || timestamp.nil?
> > -              record['time'] = timestamp
> > -            end
> > +          end
> > +
> > +          # Save the timestamp if available, then clear it out to allow for
> > +          # determining whether we should parse the log or message field.
> > +          timestamp = record.key?('time') ? record['time'] : nil
> > +          record.delete('time')
> > +          # If the log is json, we want to export it as a structured log unless
> > +          # there is additional metadata that would be lost.
> > +          is_json = false
> > +          if record.length == 1 && record.key?('log')
> > +            record_json = parse_json_or_nil(record['log'])
>
> I don't think our output plugin should arbitrarily parse JSON; it's ok to
> do it in special cases but I think it's surprising and potentially
> undesirable behavior in the general case.
>
> FWIW, at some point along the line I did start to wonder if we were doing
> the right thing by baking all of this type of functionality into our output
> plugin, or whether we should be using filter plugins
> <http://docs.fluentd.org/articles/filter-plugin-overview> or something
> else. Then we can compose and configure the proper chain of filters for a
> given environment, while making the output plugin a more generic "just send
> what you get to cloud logging". I never had the time to sit down and think
> it through, but it might be worth considering.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/GoogleCloudPlatform/fluent-plugin-google-cloud/pull/77/files/eef0d044ce874f422dff0aa2e47e51373c3dca9e#r67397835>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AALxtBc_OWEXPEBZUlyUmVP7kg0sg6-zks5qMZQogaJpZM4I1yzk>
> .
>